### PR TITLE
[Improvement] completion handle aliases

### PR DESCRIPTION
**Pull request by camilledejoye**
Hi,

That's an attempt to deal with aliases from the import table when completing classes (see https://github.com/phpactor/phpactor/issues/1121).
I tried to not keep it as clean as possible but it still kind of a POC to see how you feel about the idea so don't hesitate: the more critics the better :smile: 

I decided to put it on the `ChainTolerantCompletor` because we need the parser to get the import table and I wanted it to work for all class completors without having to deal with them one by one.

The new `DoctrineAnnotationCompletor` is a bit special so I had to extract the logic into a trait.
As I said in the commit message I did it to because it feels a right use case but if you prefer an helper class with static methods it's also possible of course.

I've mainly used the automated test for now, just one or two manual test in a demo project to make sure it works.
I will update my work environment on Monday so that I can test it in a more realistic code base.

### CODE REVIEW

File: lib/Bridge/TolerantParser/ChainTolerantCompletor.php [hunk 1]
```diff
@@ -53,9 +55,16 @@ public function complete(TextDocument $source, ByteOffset $byteOffset): Generato
...
             $suggestions = $tolerantCompletor->complete($completionNode, $source, $byteOffset);
 
-            yield from $suggestions;
+            foreach ($suggestions as $suggestion) {
+                // Trick to avoid any BC break when converting to an array
+                // https://www.php.net/manual/fr/language.generators.syntax.php#control-structures.yield.from
+                foreach ($this->resolveAliasSuggestions($importTable, $suggestion) as $resolvedSuggestion) {
+                    yield $resolvedSuggestion;
+                }
+            }
```
Comment (dantleech) [inline lib/Bridge/TolerantParser/ChainTolerantCompletor.php:67]: could this be a new decorator rather than modifying this class?

File: lib/Bridge/TolerantParser/ResolveAliasSuggestionsTrait.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace Phpactor\Completion\Bridge\TolerantParser;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\ResolvedName;
+use Phpactor\Completion\Core\Suggestion;
+
+trait ResolveAliasSuggestionsTrait
```
Comment (dantleech) [inline lib/Bridge/TolerantParser/ResolveAliasSuggestionsTrait.php:9]: would it make sense to inject this rather than using a trait?

File: lib/Bridge/TolerantParser/ChainTolerantCompletor.php (hunk 1 above)
Comment (camilledejoye) [reply lib/Bridge/TolerantParser/ChainTolerantCompletor.php:67]: I like the idea but I'm not sure it would allow us to deal with every possible case.
For example the `DoctrineAnnotationCompletor` also needs to update the `snippet` of each suggestion.

I'll think about it and gave it a try, if there is a way to update the snippet in the same way for all classes then yes and it will solve your next review too.

File: lib/Bridge/TolerantParser/ResolveAliasSuggestionsTrait.php (hunk 2 above)
Comment (camilledejoye) [reply lib/Bridge/TolerantParser/ResolveAliasSuggestionsTrait.php:9]: That's why I tried to explain in the commit message.
I'm not sure it make sense to have a service for that since it simply provide methods to use.

But with your previous comment about the decorator I think it fixes the question.
If we can't transform it into a decorator then I might have a preference for a simple helper class with static methods

**Comment by dantleech** [comment]
Instead of suggesting on `En<> => ORM\Entity` would it be an idea to instead complete on `ORM\\{Entity,Column,Foobar}` ? This also makes annotations discoverable ?

**Comment by camilledejoye** [comment]
Yes, that was the first item in https://github.com/phpactor/phpactor/issues/1121
I just think it might be a bit harder right now because of how the name lookup works but I would definitely like to add this at some point.

**Comment by dantleech** [comment]
Searching by FQN is trivial to add fortunately :relaxed: 

image

https://github.com/phpactor/indexer-extension/pull/26

**Comment by dantleech** [comment]
Given the above is now merged -- would it make sense to do _namespace_ and _import name_ completion instead of this? 

- *FQN completion*:
  - Can be clearly applied (e.g. only when the node is already a partially qualified name `ORM\\<>`)
  - Can help with discovery
  - Requiring some specification helps (e.g. if I search for `Before<>` with PHPBench + Behat + etc installed the list is quite extensive)
- *Import name completion*: e.g. `use Doctrine\Annotations;` `Anno<> => Annotations\\`

**Comment by camilledejoye** [comment]
For me this is two complementary features.
The one in this PR allows to discover aliases that you don't even realize are used and it's more convenient.

With only the search by relative qualified name we still need to type the alias like `ORM\`.
With Symfonys validator it could be really painful to always have to retype the prefix when you could simply do `NotBl<>` and have it completed properly given the import table.

While the improvement you've done on the indexer will indeed grant us the possibility to discover available names, which is always useful in new/large code base.
It does not allow to discover existing aliases, often at work I want to import a class but someone already did but used an alias for some reason. When typing the "correct" class name I'm not aware of this alias: this feature will show me a different name letting me know about the alias.

I grant you the last one is really an edge case :) 

Bottom line I think we can keep this one, with upgrade if necessary, and create another one to complete relative qualified name.

**Comment by dantleech** [comment]
> you could simply do NotBl<> and have it completed properly given the import table.

image

Maybe it should be prioritized?

**Comment by dantleech** [comment]
But that can be a separate thing -- we should add:

```
$container->register(CompletorClass::class, function () {}, [ CompletionExtension::COMPLETOR => [ 'priority' => 100]])
```

in the completion extension

File: lib/Bridge/TolerantParser/ResolveAliasSuggestionCompletor.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+     *
+     * @return Suggestion[]
+     */
+    private function resolveAliasSuggestions(array $importTable, Suggestion $suggestion): array
+    {
+        if (Suggestion::TYPE_CLASS !== $suggestion->type()) {
+            return [$suggestion];
+        }
+
+        $suggestionFqcn = $suggestion->classImport();
```
Comment (dantleech) [inline lib/Bridge/TolerantParser/ResolveAliasSuggestionCompletor.php:89]: you should use `nameImport` instead (`classImport` is deprecated) maybe this also helps it work with functions?

Comment (camilledejoye) [reply lib/Bridge/TolerantParser/ResolveAliasSuggestionCompletor.php:89]: I realized it when doing the feature but forgot to check what I had already done, I will fix it.
I'm not sure to understand the question about the functions, for now I wanted to keep it simple.
But indeed I had in mind to make it work with functions as well at some point, that's way I don't mention "class" in the name of the decorator.

**Comment by camilledejoye** [comment]
I updated with the implementation we discussed earlier.
I will create a PR right away for the `completion-worse-extension` so that you can test it without figuring out what to register as a service.
Works for me, I will gave it a try at work as much as possible to have a better feeling about it next week-end ;)
